### PR TITLE
Combine two VUPS_w2b into VUPS_x2c

### PR DESCRIPTION
**Pull request by abhinay-anubola**
Replaced v32acc32 sups and v16acc64 lups logic with builtins.

### CODE REVIEW

File: llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp [hunk 1]
```diff
@@ -266,6 +267,22 @@ bool combineINSERT_SUBREG(MachineBasicBlock &MBB) {
...
 }
 
+bool combineVUPS(MachineBasicBlock &MBB) {
+  MachineRegisterInfo &MRI = MBB.getParent()->getRegInfo();
+  const TargetInstrInfo &TII = *MBB.getParent()->getSubtarget().getInstrInfo();
+  auto *AIETII = static_cast<const AIEBaseInstrInfo *>(&TII);
+  bool Changed = false;
+  for (MachineInstr &MI : make_early_inc_range(reverse(MBB))) {
+    if (MI.getOpcode() != TargetOpcode::REG_SEQUENCE ||
+        MI.getNumOperands() != 5)
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp]: Can you fold `MI.getNumOperands() != 5` into the target specific implementation? I think the only requirement for `combineLoHiMI` should be that MI is a REG_SEQUENCE

File: llvm/lib/Target/AIE/AIE2InstrInfo.cpp [hunk 2]
```diff
@@ -1305,3 +1305,57 @@ bool AIE2InstrInfo::canHoistCheapInst(const MachineInstr &MI) const {
...
+      break;
+    case AIE2::VUPS_S64_S32_mv_ups_w2b:
+      CombinedOpCode = AIE2::VUPS_S64_S32_mv_ups_x2c;
+      break;
+    default:
+      return false;
+    }
+    Register Dst = MI.getOperand(0).getReg();
+    Register Src = MRI.createVirtualRegister(&AIE2::VEC512RegClass);
+    Register Shift = HiMI->getOperand(2).getReg();
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIE2InstrInfo.cpp]: We should also verify that the Shift amount is the same

File: llvm/lib/Target/AIE/AIE2InstrInfo.cpp [hunk 3]
```diff
@@ -1305,3 +1305,57 @@ bool AIE2InstrInfo::canHoistCheapInst(const MachineInstr &MI) const {
          AIE2::eSRegClass.hasSubClassEq(
              MRI.getRegClass(MI.getOperand(0).getReg()));
 }
+
+bool AIE2InstrInfo::combineLoHiMI(MachineInstr &MI,
+                                  MachineRegisterInfo &MRI) const {
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIE2InstrInfo.cpp]: I feel a lot of the machinery here could be share between AIE versions. Could you maybe move the MI creation/manipulation to the PostSelectOptimize pass? Then I would suggest only implementing an interface like below for each AIEInstrInfo version. E.g. for AIE2:

```
struct LoHiCombineInfo {
  unsigned CombinedOpcode;
  unsigned OpIdxToAggregate;
};

std::optional<LoHiCombineInfo> AIE2InstrInfo::getCombinedLoHiOpcode(LoMI, HiMI) {
  unsigned LoOpcode = LoMI.getOpcode();
  unsigned HIOpcode = HiMI.getOpcode();
  if (LoOpcode != HiOpcode)
    return {};

  switch (LoOpcode) {
  case AIE2::VUPS_S32_D16_mv_ups_w2b:
    if (LoMI.getOperand(2) == HiMI.getOperand(2))
      return LoHiCombineInfo{VUPS_S32_D16_mv_ups_x2c, /*OpIdxToAggregate=*/1}
    break;
  case ...
  }
  return {}
}

Comment (andcarminati) [reply llvm/lib/Target/AIE/AIE2InstrInfo.cpp]: Maybe `LoHiCombineInfo` could also include register classes for input and output.....

**Comment by konstantinschwarz** [comment]
Do we see this missing combine in real world code?
I think we discovered a missing ,combine in aiev2_ups.h: https://github.com/Xilinx/llvm-aie/blob/aie-public/clang/lib/Headers/aiev2_ups.h#L115
Might be easier to just use the proper builtin there?
@abnikant has more details

**Comment by gbossu** [comment]
> Do we see this missing combine in real world code? I think we discovered a missing ,combine in aiev2_ups.h: https://github.com/Xilinx/llvm-aie/blob/aie-public/clang/lib/Headers/aiev2_ups.h#L115 Might be easier to just use the proper builtin there? @abnikant has more details

I spotted that missing combine in `SigmoidTemplated_int8`, maybe other benchmarks as well. If we can solve it at the api level, I agree this is a good enough fix.

**Review by gbossu — COMMENTED** [review]
LGTM, but maybe you could sqush the two commits together? I think it's always better to have the change and the corresponding tests in the same commit.

**Review by gbossu — APPROVED** [review]

**Pull request merged**